In [6]:
from pathlib import Path
import sys

print("Python:", sys.executable)
print("현재 위치:", Path.cwd())

Python: c:\dev\llm-data-analysis-study\.venv\Scripts\python.exe
현재 위치: c:\dev\llm-data-analysis-study\chapter03


In [7]:
from pathlib import Path
import pandas as pd

PROJECT_ROOT = Path.cwd()

if not (PROJECT_ROOT / "data" / "raw").exists():
    PROJECT_ROOT = PROJECT_ROOT.parent

DATA_DIR = PROJECT_ROOT / "data" / "raw"

customers = pd.read_csv(DATA_DIR / "customers.csv")
products = pd.read_csv(DATA_DIR / "products.csv")
orders = pd.read_csv(DATA_DIR / "orders.csv")
order_items = pd.read_csv(DATA_DIR / "order_items.csv")

datasets = {
    "customers": customers,
    "products": products,
    "orders": orders,
    "order_items": order_items,
}

print("PROJECT_ROOT:", PROJECT_ROOT)
print("DATA_DIR:", DATA_DIR)

for name, df in datasets.items():
    print(name, df.shape)

PROJECT_ROOT: c:\dev\llm-data-analysis-study
DATA_DIR: c:\dev\llm-data-analysis-study\data\raw
customers (150, 6)
products (100, 4)
orders (300, 5)
order_items (764, 5)


In [8]:
for name, df in datasets.items():
    print(f"\n===== {name} =====")
    print("shape:", df.shape)
    print("columns:", list(df.columns))
    print("dtypes:")
    print(df.dtypes)


===== customers =====
shape: (150, 6)
columns: ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
dtypes:
customer_id    int64
name             str
gender           str
age            int64
city             str
signup_date      str
dtype: object

===== products =====
shape: (100, 4)
columns: ['product_id', 'product_name', 'category', 'price']
dtypes:
product_id      int64
product_name      str
category          str
price           int64
dtype: object

===== orders =====
shape: (300, 5)
columns: ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
dtypes:
order_id          int64
customer_id       int64
order_date          str
payment_method      str
order_status        str
dtype: object

===== order_items =====
shape: (764, 5)
columns: ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']
dtypes:
order_item_id    int64
order_id         int64
product_id       int64
quantity         int64
unit_price       int64
dtype: object


In [9]:
customers.head()
products.head()
orders.head()
order_items.head()

,order_item_id,order_id,product_id,quantity,unit_price
0,1,1,100,3,102000
1,2,1,87,5,25000
2,3,1,7,3,142000
3,4,1,9,3,193000
4,5,2,72,4,189000


In [10]:
for name, df in datasets.items():
    print(f"{name}: shape={df.shape}")
    print("columns:", list(df.columns))
    print("dtypes:", df.dtypes.astype(str).to_dict())
    print()

customers: shape=(150, 6)
columns: ['customer_id', 'name', 'gender', 'age', 'city', 'signup_date']
dtypes: {'customer_id': 'int64', 'name': 'str', 'gender': 'str', 'age': 'int64', 'city': 'str', 'signup_date': 'str'}

products: shape=(100, 4)
columns: ['product_id', 'product_name', 'category', 'price']
dtypes: {'product_id': 'int64', 'product_name': 'str', 'category': 'str', 'price': 'int64'}

orders: shape=(300, 5)
columns: ['order_id', 'customer_id', 'order_date', 'payment_method', 'order_status']
dtypes: {'order_id': 'int64', 'customer_id': 'int64', 'order_date': 'str', 'payment_method': 'str', 'order_status': 'str'}

order_items: shape=(764, 5)
columns: ['order_item_id', 'order_id', 'product_id', 'quantity', 'unit_price']
dtypes: {'order_item_id': 'int64', 'order_id': 'int64', 'product_id': 'int64', 'quantity': 'int64', 'unit_price': 'int64'}



## STEP 4 데이터 구조 확인

### 결과 관찰

customers는 150행 6열, products는 100행 4열, orders는 300행 5열, order_items는 764행 5열로 확인되었다.

각 데이터에는 고객, 상품, 주문, 주문 상세 정보를 구분할 수 있는 ID 컬럼이 있었으며, age, price, quantity, unit_price와 같은 값은 숫자형으로 읽혔다.

반면 customers의 signup_date와 orders의 order_date는 날짜처럼 보이지만 현재 문자열 형태로 읽혀 있었다.

### 나의 해석과 판단

각 데이터셋의 기본 구조와 주요 컬럼을 확인할 수 있었고, 여러 CSV를 서로 연결할 때 사용할 수 있는 ID 컬럼도 확인했다.

날짜 컬럼은 현재 문자열이므로 날짜 범위 확인이나 날짜 기준 분석을 진행하기 전에 날짜 타입으로 변환할 필요가 있다고 판단했다.

### 업무·분석적 의미

분석을 시작하기 전에 데이터의 행과 열 개수, 컬럼명, 데이터 타입을 확인하면 잘못된 컬럼을 사용하거나 데이터 타입을 잘못 해석하는 문제를 줄일 수 있다.

특히 여러 데이터를 연결해서 분석할 경우 각 ID 컬럼이 어떤 역할을 하는지 미리 파악하는 것이 중요하다.

### 한계와 추가 확인 사항

현재는 데이터 구조와 일부 값만 확인했기 때문에 데이터 전체의 품질이 정상이라고 판단할 수는 없다.

다음 단계에서 결측값, 중복값, 주요 ID의 고유성 등을 추가로 확인해야 한다.

![데이터 구조 확인](images/step01_structure.png)

In [11]:
for name, df in datasets.items():
    print(f"\n===== {name} 결측치 =====")
    print(df.isna().sum())

print("\n===== 전체 행 중복 =====")
for name, df in datasets.items():
    print(name, df.duplicated().sum())


===== customers 결측치 =====
customer_id    0
name           0
gender         0
age            0
city           0
signup_date    0
dtype: int64

===== products 결측치 =====
product_id      0
product_name    0
category        0
price           0
dtype: int64

===== orders 결측치 =====
order_id          0
customer_id       0
order_date        0
payment_method    0
order_status      0
dtype: int64

===== order_items 결측치 =====
order_item_id    0
order_id         0
product_id       0
quantity         0
unit_price       0
dtype: int64

===== 전체 행 중복 =====
customers 0
products 0
orders 0
order_items 0


In [12]:
key_columns = {
    "customers": "customer_id",
    "products": "product_id",
    "orders": "order_id",
    "order_items": "order_item_id",
}

for name, key in key_columns.items():
    df = datasets[name]

    print(f"\n===== {name}.{key} =====")
    print("결측:", df[key].isna().sum())
    print("중복:", df[key].duplicated().sum())



===== customers.customer_id =====
결측: 0
중복: 0

===== products.product_id =====
결측: 0
중복: 0

===== orders.order_id =====
결측: 0
중복: 0

===== order_items.order_item_id =====
결측: 0
중복: 0


## STEP 5 결측치 및 중복 확인

### 결과 관찰

customers, products, orders, order_items의 각 컬럼에서 결측치는 발견되지 않았다.

또한 네 데이터셋 모두 전체 행 중복은 0건으로 확인되었다.

주요 ID 컬럼인 customer_id, product_id, order_id, order_item_id도 모두 결측 0건, 중복 0건으로 확인되었다.

### 나의 해석과 판단

현재 샘플 데이터에서는 주요 ID의 고유성이 유지되고 있어 각 테이블의 식별자로 사용할 수 있을 가능성이 높다고 판단했다.

전체 행 중복도 발견되지 않았기 때문에 동일한 데이터가 완전히 중복되어 저장된 문제는 현재 확인되지 않았다.

### 업무·분석적 의미

주요 ID에 결측이나 중복이 있으면 이후 여러 데이터를 연결할 때 행이 누락되거나 중복 집계가 발생할 수 있다.

따라서 분석 전에 주요 식별자의 품질을 먼저 확인하는 것이 중요하다.

### 한계와 추가 확인 사항

현재는 주요 ID와 전체 행 중복 여부를 확인한 단계이므로, 연결용 외래키가 실제 부모 테이블에 존재하는지까지 확인한 것은 아니다.

다음 단계에서 각 CSV 사이의 키 관계가 정상적으로 연결되는지도 추가로 확인해야 한다.

![결측 및 중복 확인](images/step02_quality.png)


In [13]:
print("===== 숫자형 요약 =====")
print("\n[customers.age]")
print(customers[["age"]].describe())

print("\n[products.price]")
print(products[["price"]].describe())

print("\n[order_items quantity / unit_price]")
print(order_items[["quantity", "unit_price"]].describe())

===== 숫자형 요약 =====

[customers.age]
              age
count  150.000000
mean    42.086667
std     15.613166
min     19.000000
25%     29.000000
50%     40.000000
75%     57.000000
max     69.000000

[products.price]
               price
count     100.000000
mean   110040.000000
std     56433.910574
min      5000.000000
25%     65750.000000
50%    112000.000000
75%    161000.000000
max    200000.000000

[order_items quantity / unit_price]
         quantity     unit_price
count  764.000000     764.000000
mean     3.053665  108561.518325
std      1.410873   56996.770604
min      1.000000    5000.000000
25%      2.000000   62000.000000
50%      3.000000  111000.000000
75%      4.000000  161250.000000
max      5.000000  200000.000000


In [14]:
print("===== 범주형 값 확인 =====")

print("\n[customers.city]")
print(customers["city"].value_counts(dropna=False))

print("\n[products.category]")
print(products["category"].value_counts(dropna=False))

print("\n[orders.order_status]")
print(orders["order_status"].value_counts(dropna=False))

===== 범주형 값 확인 =====

[customers.city]
city
성남    21
광주    17
부산    16
대구    15
서울    15
울산    14
인천    14
대전    14
수원    13
고양    11
Name: count, dtype: int64

[products.category]
category
스포츠     19
전자기기    17
생활용품    16
뷰티      16
도서      14
패션      11
식품       7
Name: count, dtype: int64

[orders.order_status]
order_status
completed    184
cancelled     64
refunded      52
Name: count, dtype: int64


In [15]:
orders["order_date"] = pd.to_datetime(
    orders["order_date"],
    errors="coerce",
)

print("변환 실패 건수:", orders["order_date"].isna().sum())
print("가장 빠른 날짜:", orders["order_date"].min())
print("가장 최근 날짜:", orders["order_date"].max())

변환 실패 건수: 0
가장 빠른 날짜: 2025-09-03 00:00:00
가장 최근 날짜: 2026-09-02 00:00:00


## STEP 6 숫자형·범주형·날짜 데이터 확인

### 결과 관찰

customers의 age는 최소 19세, 최대 69세였고 평균은 약 42.09세, 중앙값은 40세였다.

products의 price는 최소 5,000원, 최대 200,000원이었으며 평균은 약 110,040원, 중앙값은 112,000원이었다.

order_items의 quantity는 중앙값이 3개였고 최대 5개까지 확인되었다. unit_price는 중앙값이 111,000원이었고 최대 200,000원이었다.

범주형 데이터에서는 customers.city 중 성남이 21건으로 가장 많았고, products.category에서는 스포츠가 19건으로 가장 많았다.

orders.order_status는 completed 184건, cancelled 64건, refunded 52건으로 확인되었다.

order_date를 날짜 타입으로 변환한 결과 변환 실패는 0건이었고, 가장 빠른 주문일은 2025-09-03, 가장 최근 주문일은 2026-09-02였다.

### 나의 해석과 판단

현재 확인한 숫자형 데이터의 최소값과 최대값은 눈에 띄게 비정상적으로 보이는 값은 없었다.

범주형 데이터는 값별 빈도 차이가 있었지만, 현재 단계에서는 단순한 편중인지 실제 데이터 특성인지 추가 확인이 필요하다고 판단했다.

order_date는 문자열 상태로 읽혔지만 날짜 타입 변환에 실패한 값이 없었기 때문에 현재 샘플 데이터에서는 날짜 형식이 비교적 일관되게 저장되어 있다고 판단했다.

### 업무·분석적 의미

숫자형 데이터의 범위를 미리 확인하면 잘못 입력된 값이나 비정상적인 값이 존재하는지 빠르게 확인할 수 있다.

범주형 데이터의 분포를 확인하면 특정 지역, 상품군, 주문 상태에 데이터가 지나치게 몰려 있는지 파악할 수 있다.

또한 날짜 데이터를 실제 날짜 타입으로 변환해 두면 이후 기간별 주문 분석이나 시계열 분석을 수행하기 쉬워진다.

### 한계와 추가 확인 사항

현재는 기본적인 범위와 분포만 확인했기 때문에 실제 업무 규칙에 맞는 값인지까지 판단할 수는 없다.

이상값으로 보이는 값이 발견될 경우 임의로 수정하거나 삭제하기보다 실제 업무에서 가능한 값인지 추가 정보와 함께 확인할 필요가 있다.

![숫자형·범주형·날짜 데이터 확인](images/step03_distribution.png)


In [16]:
missing_customer_ids = orders.loc[
    ~orders["customer_id"].isin(customers["customer_id"]),
    "customer_id"
].unique()

missing_order_ids = order_items.loc[
    ~order_items["order_id"].isin(orders["order_id"]),
    "order_id"
].unique()

missing_product_ids = order_items.loc[
    ~order_items["product_id"].isin(products["product_id"]),
    "product_id"
].unique()

print("customers에 없는 customer_id:", len(missing_customer_ids))
print("orders에 없는 order_id:", len(missing_order_ids))
print("products에 없는 product_id:", len(missing_product_ids))


customers에 없는 customer_id: 0
orders에 없는 order_id: 0
products에 없는 product_id: 0


## STEP 7 CSV 간 키 관계 검증

### 결과 관찰

orders의 customer_id 중 customers에 존재하지 않는 값은 0개였다.

order_items의 order_id 중 orders에 존재하지 않는 값은 0개였고, order_items의 product_id 중 products에 존재하지 않는 값도 0개였다.

### 나의 해석과 판단

현재 샘플 데이터에서는 자식 테이블에 있는 연결용 ID가 모두 부모 테이블에 존재하고 있어 기본적인 키 관계는 정상적으로 연결되어 있다고 판단했다.

즉, 현재 확인한 범위에서는 orphan key가 발견되지 않았다.

### 업무·분석적 의미

테이블 간 연결용 키가 정상적으로 존재해야 이후 merge나 join을 수행할 때 데이터가 누락되거나 잘못 연결되는 문제를 줄일 수 있다.

여러 CSV를 함께 분석할 경우 개별 테이블의 ID 고유성뿐 아니라 테이블 사이의 연결 관계도 함께 확인하는 것이 중요하다.

### 한계와 추가 확인 사항

현재는 연결 대상 ID가 존재하는지만 확인했기 때문에 각 데이터 값 자체의 정확성이나 실제 업무 규칙까지 검증한 것은 아니다.

추후 실제 운영 데이터에서는 키 생성 규칙이나 데이터 변경 이력 등도 추가로 확인할 필요가 있다.

![CSV 간 키 관계 검증](images/step04_relationship.png)


## STEP 8 LLM을 활용한 추가 데이터 품질 점검

### LLM에 제공한 Safe Context
원본 고객 데이터나 개인정보는 제공하지 않고, 각 데이터셋의 shape, 컬럼명, 데이터 타입, 결측 및 중복 여부, 숫자형 범위, 범주형 분포, 키 관계 검증 결과만 제공했다.

### LLM이 제안한 추가 점검 항목
LLM은 다음과 같은 항목을 추가로 확인할 것을 제안했다.

- signup_date 날짜 변환 가능 여부
- 숫자형 값의 이상 범위 확인
- 범주형 값의 표기 일관성
- 주문 상태별 분포 확인
- products.price와 order_items.unit_price의 관계 확인
- 고객 가입일과 주문일의 관계 확인
- 취소 및 환불 처리 기준 확인

### 실제 데이터에서 확인 가능한 내용
현재 데이터만으로 signup_date의 날짜 변환 여부, 숫자형 값의 범위, 범주형 값의 종류와 빈도, products.price와 order_items.unit_price의 차이 등을 추가로 확인할 수 있다.

반면 나이와 가격이 실제 업무 규칙상 정상 범위인지, 취소 및 환불 상태의 정확한 업무 기준이 무엇인지는 현재 데이터만으로 단정하기 어렵다.

### 채택 / 수정 / 보류한 내용
- 채택: signup_date 날짜 변환 여부, 숫자형 값 범위, 범주형 값 표기 일관성 확인
- 수정: 가격 차이는 단순 오류로 판단하지 않고 할인이나 가격 변경 가능성을 함께 고려
- 보류: 취소 및 환불 처리 기준, 고객 연령 기준 등 업무 규칙이 필요한 항목

### 나의 판단
LLM의 제안은 추가 점검 항목을 찾는 데 도움이 되었지만, 모든 제안을 바로 문제로 판단할 수는 없었다.

특히 데이터에서 직접 확인할 수 있는 사실과 실제 업무 정의가 필요한 내용을 구분해야 한다고 판단했다.

이번 Chapter에서는 데이터를 바로 수정하기보다 현재 상태를 관찰하고 기록하는 것이 목적이므로, 이상해 보이는 값이 있더라도 원인을 확인하기 전에는 임의로 삭제하거나 수정하지 않는 것이 적절하다고 판단했다.

![LLM 제안과 검증](images/step05_llm.png)

## Chapter 03 최종 정리

### 데이터의 첫인상 3가지
1. customers, products, orders, order_items의 주요 ID는 현재 샘플 데이터에서 결측과 중복 없이 고유성을 유지하고 있었다.
2. signup_date와 order_date는 처음에는 문자열 형태로 읽혔으며, order_date는 날짜 타입으로 변환했을 때 실패값이 없었다.
3. CSV 간 키 관계를 확인한 결과 orders.customer_id, order_items.order_id, order_items.product_id는 모두 연결 대상 부모 테이블에 존재했다.

### 다음 Chapter 전에 확인하거나 처리해야 할 항목
1. signup_date도 날짜 타입으로 변환 가능한지 확인한다.
2. 숫자형 값과 범주형 값 중 실제 업무 기준상 이상한 값이 있는지 추가 확인한다.
3. products.price와 order_items.unit_price의 관계와 취소·환불 주문 처리 기준을 확인한다.

### 현재 데이터만으로 단정할 수 없는 것
현재 데이터만으로는 나이와 가격 범위가 실제 업무 규칙상 정상인지, 취소 및 환불 상태가 어떤 기준으로 처리되는지까지 판단할 수 없다.

또한 가격 차이가 발견되더라도 할인, 행사, 가격 변경 등 업무상의 이유가 있을 수 있으므로 추가 정보 없이 오류라고 단정하면 안 된다.